<?xml version="1.0"?>
<robot name="mrobot" xmlns:xacro="http://www.ros.org/wiki/xacro">

    <xacro:property name="M_PI" value="3.1415926" />
    <xacro:property name="mass" value="1" />
    <xacro:property name="base_link_length" value="0.5" />
    <xacro:property name="base_link_width" value="0.7" />
    <xacro:property name="base_link_height" value="0.1" />
    <xacro:property name="wheel_radius" value="0.08" />
    <xacro:property name="wheel_length" value="0.04" />
    <xacro:property name="wheel_joint_x" value="0.27" />
    <xacro:property name="wheel_joint_y" value="0.27" />
    <xacro:property name="wheel_joint_z" value="0.08" />

    <!-- 导入所有gazebo标签及插件 -->
    <xacro:include filename="$(find agv_robot)/urdf/mrobot.gazebo" />
    <!-- 导入颜色 -->
    <xacro:include filename="$(find agv_robot)/urdf/materials.xacro" />
  
    <xacro:macro name="default_inertial" params="mass">
        <inertial>
            <origin xyz="0 0 0" rpt="0 0 0" />
            <mass value="${mass}" />
            <inertia ixx="${mass/10}" ixy="0" ixz="0"
                     iyy="${mass/10}" iyz="0"
                     izz="${mass/10}" />
        </inertial>
    </xacro:macro>

    <link name="world"/>
    <joint name="fixed" type="fixed">
        <parent link="world"/>
        <child link="base_link"/>
    </joint>

    <!--base_link-->
    <link name="base_link" >
        <collision>
            <origin xyz="0 0 ${wheel_radius*2 + base_link_height/2}" rpy="0 0 0" />
            <geometry>
                <box size="${base_link_length}
                           ${base_link_width} 
                           ${base_link_height}" />
            </geometry>
        </collision>

        <visual>
            <origin xyz="0 0 ${wheel_radius*2 + base_link_height/2}" rpy="0 0 0" />
            <geometry>
                <box size="${base_link_length}
                           ${base_link_width} 
                           ${base_link_height}" />
            </geometry>
            <material name="orange">
                <color rgba="${255/255} ${108/255} ${10/255} 1.0"/>
            </material>
        </visual>
        <xacro:default_inertial mass="${mass}" />
    </link>

  <gazebo reference="base_link">
    <material>Gazebo/orange</material>
  </gazebo>

    <joint name="joint1" type="continuous">
        <parent link="base_link" />
        <child link="wheel1_link" />
        <origin xyz="0 0 0"
                rpy="0 0 0" /> 
        <axis xyz="0 1 0"/>
    </joint>
    
    <link name="wheel1_link">
        <collision>
            <origin xyz="${wheel_joint_x+wheel_length/2}
                     ${wheel_joint_y}
                     ${wheel_joint_z}" rpy="0 ${M_PI/2} 0" />
            <geometry>
                <cylinder length="${wheel_length}" 
                          radius="${wheel_radius}" />
            </geometry>
        </collision>

        <visual>
            <origin xyz="${wheel_joint_x+wheel_length/2}
                     ${wheel_joint_y}
                     ${wheel_joint_z}"
                    rpy="0 ${M_PI/2} 0" />
            <geometry>
                <cylinder length="${wheel_length}" 
                          radius="${wheel_radius}" />
            </geometry>
            <material name="black">
   	 	<color rgba="0.0 0.0 0.0 1.0"/>
  	    </material>
         </visual>
         <xacro:default_inertial mass="${mass}" />
    </link>

  <gazebo reference="wheel1_link">
    <material>Gazebo/black</material>
  </gazebo>

    <joint name="joint2" type="continuous">
        <parent link="base_link" />
        <child link="wheel2_link" />
        <origin xyz="0 0 0"
                rpy="0 0 0" /> 
        <axis xyz="0 1 0"/>
    </joint>

    <link name="wheel2_link">
        <collision>
            <origin xyz="-${wheel_joint_x+wheel_length/2}
                     ${wheel_joint_y}
                     ${wheel_joint_z}" rpy="0 ${M_PI/2} 0" />
            <geometry>
                <cylinder length="${wheel_length}" 
                          radius="${wheel_radius}" />
            </geometry>
        </collision>

        <visual>
            <origin xyz="-${wheel_joint_x+wheel_length/2}
                     ${wheel_joint_y}
                     ${wheel_joint_z}"
                    rpy="0 ${M_PI/2} 0" />
            <geometry>
                <cylinder length="${wheel_length}" 
                          radius="${wheel_radius}" />
            </geometry>
            <material name="black">
    		<color rgba="0.0 0.0 0.0 1.0"/>
            </material>
         </visual>
         <xacro:default_inertial mass="${mass}" />
    </link>

  <gazebo reference="wheel2_link">
    <material>Gazebo/black</material>
  </gazebo>

    <joint name="joint3" type="continuous">
        <parent link="base_link" />
        <child link="wheel3_link" />
        <origin xyz="0 0 0"
                rpy="0 0 0" /> 
        <axis xyz="0 1 0"/>
    </joint>

    <link name="wheel3_link">
        <collision>
            <origin xyz="-${wheel_joint_x+wheel_length/2}
                     -${wheel_joint_y}
                     ${wheel_joint_z}" rpy="0 ${M_PI/2} 0" />
            <geometry>
                <cylinder length="${wheel_length}" 
                          radius="${wheel_radius}" />
            </geometry>
        </collision>

        <visual>
            <origin xyz="-${wheel_joint_x+wheel_length/2}
                     -${wheel_joint_y}
                     ${wheel_joint_z}"
                    rpy="0 ${M_PI/2} 0" />
            <geometry>
                <cylinder length="${wheel_length}" 
                          radius="${wheel_radius}" />
            </geometry>
            <material name="black">
    		<color rgba="0.0 0.0 0.0 1.0"/>
  	    </material>
         </visual>
         <xacro:default_inertial mass="${mass}" />
    </link>

<gazebo reference="wheel3_link">
    <material>Gazebo/black</material>
  </gazebo>

    <joint name="joint4" type="continuous">
        <parent link="base_link" />
        <child link="wheel4_link" />
        <origin xyz="0 0 0"
                rpy="0 0 0" /> 
        <axis xyz="0 1 0"/>
    </joint>

    <link name="wheel4_link">
        <collision>
            <origin xyz="${wheel_joint_x+wheel_length/2}
                     -${wheel_joint_y}
                     ${wheel_joint_z}" rpy="0 ${M_PI/2} 0" />
            <geometry>
                <cylinder length="${wheel_length}" 
                          radius="${wheel_radius}" />
            </geometry>
        </collision>

        <visual>
            <origin xyz="${wheel_joint_x+wheel_length/2}
                     -${wheel_joint_y}
                     ${wheel_joint_z}"
                    rpy="0 ${M_PI/2} 0" />
            <geometry>
                <cylinder length="${wheel_length}" 
                          radius="${wheel_radius}" />
            </geometry>
            <material name="black">
    		<color rgba="0.0 0.0 0.0 1.0"/>
  	    </material>
         </visual>
         <xacro:default_inertial mass="${mass}" />
    </link>

<gazebo reference="wheel4_link">
    <material>Gazebo/black</material>
  </gazebo>

    <transmission name="tran1">
        <type>transmission_interface/SimpleTransmission</type>
        <joint name="joint1">
            <hardwareInterface>hardware_interface/EffortJointInterface</hardwareInterface>
        </joint>
        <actuator name="motor1">
            <hardwareInterface>hardware_interface/EffortJointInterface</hardwareInterface>
            <mechanicalReduction>1</mechanicalReduction>
        </actuator>
    </transmission>

    <transmission name="tran2">
        <type>transmission_interface/SimpleTransmission</type>
        <joint name="joint2">
            <hardwareInterface>hardware_interface/EffortJointInterface</hardwareInterface>
        </joint>
        <actuator name="motor2">
            <hardwareInterface>hardware_interface/EffortJointInterface</hardwareInterface>
            <mechanicalReduction>1</mechanicalReduction>
        </actuator>
    </transmission>

    <transmission name="tran3">
        <type>transmission_interface/SimpleTransmission</type>
        <joint name="joint3">
            <hardwareInterface>hardware_interface/EffortJointInterface</hardwareInterface>
        </joint>
        <actuator name="motor3">
            <hardwareInterface>hardware_interface/EffortJointInterface</hardwareInterface>
            <mechanicalReduction>1</mechanicalReduction>
        </actuator>
    </transmission>

    <transmission name="tran4">
        <type>transmission_interface/SimpleTransmission</type>
        <joint name="joint4">
            <hardwareInterface>hardware_interface/EffortJointInterface</hardwareInterface>
        </joint>
        <actuator name="motor4">
            <hardwareInterface>hardware_interface/EffortJointInterface</hardwareInterface>
            <mechanicalReduction>1</mechanicalReduction>
        </actuator>
    </transmission>
</robot>
